File: hard/Lc332/Lc332.go
package Lc332

import (
	"container/heap"
	. "github.com/xiaozhuoops/leetcode/template/heap"
)

func findItinerary(tickets [][]string) []string {
	// 1.build graph
	graph := map[string]*StringHeap{}
	for _, ticket := range tickets {
		from, to := ticket[0], ticket[1]
		if adjList, ok := graph[from]; ok {
			heap.Push(adjList, to)
		} else {
			sh := &StringHeap{}
			heap.Init(sh)
			heap.Push(sh, to)
			graph[from] = sh
		}
	}
	// fmt.Println(graph)

	// 2.dfs
	var stack []string
	var dfs func(start string)
	dfs = func(start string) {
		for graph[start] != nil && len(*graph[start]) > 0 {
			to := heap.Pop(graph[start])
			dfs(to.(string))
		}
		stack = append(stack, start)
	}
	dfs("JFK")

	ans := make([]string, len(stack))
	for i, _ := range stack {
		ans[i] = stack[len(stack)-1-i]
	}
	return ans
}

File: hard/Lc332/Lc332.py
import heapq


class Solution:
    """
    2022年4月13日
    1 dfs
    2 O(E)
    3
    """

    def findItinerary(self, tickets: List[List[str]]) -> List[str]:
        graph = {}
        for edge in tickets:
            start, end = edge[0], edge[1]
            if start not in graph:
                graph[start] = []
            # graph[start].append(end)
            heapq.heappush(graph[start], end)
        ans = []

        def dfs(start):
            while graph[start]:
                dfs(heapq.heappop())
            ans.append(start)

        dfs("JFK")
        return ans
